modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/UserCardTag.java

```diff
@@ -0,0 +1,584 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.taglib.clay.servlet.taglib;
+
+import com.liferay.frontend.taglib.clay.internal.servlet.taglib.BaseContainerTag;
+import com.liferay.frontend.taglib.clay.servlet.taglib.soy.UserCard;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.DropdownItem;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.LabelItem;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.JspWriter;
+
+/**
+ * @author Carlos Lancha
+ */
+public class UserCardTag extends BaseContainerTag {
+
+	@Override
+	public int doStartTag() throws JspException {
+		setAttributeNamespace(_ATTRIBUTE_NAMESPACE);
+
+		ThemeDisplay themeDisplay = (ThemeDisplay)request.getAttribute(
+			WebKeys.THEME_DISPLAY);
+
+		String pathThemeImages = themeDisplay.getPathThemeImages();
+
+		_spritemap = pathThemeImages.concat("/clay/icons.svg");
+
+		return super.doStartTag();
+	}
+
+	public List<DropdownItem> getActionDropdownItems() {
+		if ((_actionDropdownItems == null) && (_userCard != null)) {
+			return _userCard.getActionDropdownItems();
+		}
+
+		return _actionDropdownItems;
+	}
+
+	@Override
+	public String getCssClass() {
+		if ((super.getCssClass() == null) && (_userCard != null)) {
+			return _userCard.getCssClass();
+		}
+
+		return super.getCssClass();
+	}
+
+	/**
+	 * @deprecated As of Cavanaugh (7.4.x), with no direct replacement
+	 */
+	@Deprecated
+	@Override
+	public Map<String, String> getData() {
```

Comment:
This will not be enough to correctly set `data` attributes. You are setting them for client-side, but not for server-side rendering. Is the idea that these will never be relevant until hydration? I'm not sure.

To add them on server side, I did this:
```
public int doStartTag() throws JspException {
        Map<String, String> data = getData();

	if (Validator.isNotNull(data)) {
		for (Map.Entry<String, String> entry : data.entrySet()) {
			setDynamicAttribute(
				StringPool.BLANK, "data-" + entry.getKey(),
				entry.getValue());
		}
	}

		...
}
```

Reply:
Why do we need them in server side?

Reply:
Here is an example. One usage of my card is on creation of a new page. When you click on a card, modal opens with form to set page name. URL for modal iframe is in card data attribute. The click event is not defined in React, but in JSP. As a result, if user clicks on card before hydration is done, empty modal will open.

Reply:
Not my case here. I'd say we should only add it if needed, and even in that case, we should enforce that case to migrate to use `propsTransformer` so everything happens in the same process, so you won't need that. Makes sense?

Of course that "enforcing" takes time, I'm not saying that we should not add them if they're needed right now, but have that in mind to remove it in the future.

Reply:
Generally speaking, isn't the idea to print all HTML on server side, then leave JS for React component to fill in? `data` attribute is a part of HTML.

Reply:
From the epic:
```
Use Java to render minimal markup server-side
```
So we don't need to render ALL Html, just the minimal markup to make it look good.

i.e: `DropdownMenuTag` only renders a button on server side rendering, not the dropdown itself.